### ISSUE TITLE
7.4 Android app not responding

### ISSUE THREAD

**Issue by Beuc** [renpy/renpy#2574]
Hi,

When upgrading to Ren'Py 7.4 and building e.g. The Question, the game launches, title music can play, but the game doesn't respond to touch events so it's stuck on title screen. Eventually Android detects this and offers to kill the unresponsive app.
Tested on Android 9, Ren'Py/RAPT installed today.

**Comment by renpytom** [comment]
Hm... It works for me on Android 11. Is there any relevant output from logcat? 

Also, if you can get today's nightly when it builds, it might be worth trying it there.

**Comment by Beuc** [comment]
No logcat output after the title screen is displayed. The previous output looked fine to me but I now have a doubt about those "rejecting buffer":
```
01-06 12:00:25.687  7650  7674 I python  : Interface start took 0.15s
01-06 12:00:25.737   398  3074 D SurfaceFlinger: duplicate layer name: changing SurfaceView - org.renpy.question/org.renpy.android.PythonSDLActivity to SurfaceView - org.renpy.question/org.renpy.android.PythonSDLActivity#1
01-06 12:00:25.737  7650  7674 W Adreno-EGL: <qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE
01-06 12:00:25.739  7650  7674 D vndksupport: Loading /vendor/lib/hw/gralloc.msm8974.so from current namespace instead of sphal namespace.
01-06 12:00:25.742   398   489 D SurfaceFlinger: duplicate layer name: changing Background for -SurfaceView - org.renpy.question/org.renpy.android.PythonSDLActivity to Background for -SurfaceView - org.renpy.question/org.renpy.android.PythonSDLActivity#1
01-06 12:00:25.744  7650  7650 V SDL     : surfaceDestroyed()
01-06 12:00:25.744  7650  7650 V SDL     : nativePause()
01-06 12:00:25.776   396   583 D AudioFlinger: mixer(0xad6030c0) throttle end: throttle time(69)
01-06 12:00:26.285  7650  7650 V SDL     : surfaceCreated()
01-06 12:00:26.287  7650  7650 V SDL     : surfaceChanged()
01-06 12:00:26.287  7650  7650 V SDL     : pixel format RGBA_8888
01-06 12:00:26.299  7650  7650 V SDL     : Window size: 1920x1080
01-06 12:00:26.300  7650  7650 V SDL     : Device size: 1920x1080
01-06 12:00:26.308  7650  7650 W Adreno-EGL: <qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE
01-06 12:00:26.309  7650  7650 D vndksupport: Loading /vendor/lib/hw/gralloc.msm8974.so from current namespace instead of sphal namespace.
01-06 12:00:26.310  7650  7650 V SDL     : nativeResume()
01-06 12:00:26.454  7650  7674 I python  : Total time until interface ready: 4.14915704727s
01-06 12:00:26.485  7650  7650 V SDL     : surfaceChanged()
01-06 12:00:26.485  7650  7650 V SDL     : pixel format RGBA_8888
01-06 12:00:26.486  7650  7650 V SDL     : Window size: 1920x1008
01-06 12:00:26.486  7650  7650 V SDL     : Device size: 1920x1080
01-06 12:00:26.800   398   398 E SurfaceFlinger: [SurfaceView - org.renpy.question/org.renpy.android.PythonSDLActivity#1] rejecting buffer: bufWidth=1920, bufHeight=1080, front.active.{w=1920, h=1008}
01-06 12:00:26.817  7650  7650 V SDL     : surfaceChanged()
01-06 12:00:26.817  7650  7650 V SDL     : pixel format RGBA_8888
01-06 12:00:26.818  7650  7650 V SDL     : Window size: 1920x1080
01-06 12:00:26.818  7650  7650 V SDL     : Device size: 1920x1080
01-06 12:00:27.050   398   398 E SurfaceFlinger: [SurfaceView - org.renpy.question/org.renpy.android.PythonSDLActivity#1] rejecting buffer: bufWidth=1920, bufHeight=1008, front.active.{w=1920, h=1080}
01-06 12:00:27.228  7650  7674 I python  : Entered background.
01-06 12:00:27.228  7650  7674 I python  : Entering foreground.
```

Today's nightly is not there yet, the current one is probably impacted by #2566.
I'll have to try again later.

**Comment by renpytom** [comment]
It should be up now.

**Comment by Beuc** [comment]
Hi. The latest nightly yields the same behavior and output.
Incidentally I have randomly had a black screen instead of the title screen.

**Comment by renpytom** [comment]
Not great. What device is this?

Also, those EGL_BAD_ATTRIBUTES are concerning.

**Comment by Beuc** [comment]
It's a Galaxy S5, with basically bare-bones LineageOS.

**Comment by renpytom** [comment]
Hm... I've tried this on 4 different devices: 

- Pixel 5
- Oneplus 3
- Fire 10 HD (7th Gen)
- Google Nexus 9

With no luck repeating the problems - it works fine on all three.  I'm somewhat limited as to what I can do to support LineageOS devices, but there are still a couple of things to try. First off, let's try running it in gl2 mode, which I think should be supported, by adding 

```renpy
define config.gl2 = True
``` 

to the game. That goes through a different video path with different setting, which might work better. After that, I've just added a new callback, to let us experiment with different GL attributes. You'd use it - outside of GL2 mode with something like:

```renpy
init python:
    def set_gl_attributes():
        import pygame 

        pygame.display.gl_reset_attributes()

        pygame.display.hint("SDL_OPENGL_ES_DRIVER", "1")
        
        # Maybe comment these two lines out?
        pygame.display.gl_set_attribute(pygame.GL_CONTEXT_MAJOR_VERSION, 2);
        pygame.display.gl_set_attribute(pygame.GL_CONTEXT_MINOR_VERSION, 0);
        
        pygame.display.gl_set_attribute(pygame.GL_CONTEXT_PROFILE_MASK, pygame.GL_CONTEXT_PROFILE_ES)

        # pygame.display.gl_set_attribute(pygame.GL_SWAP_CONTROL, vsync)
        # pygame.display.gl_set_attribute(pygame.GL_ALPHA_SIZE, 8)
 
    config.set_gl_attributes = set_gl_attributes
```

This will be in this morning's nightly, but if you need it earlier let me know and I can make a special prerelease.

**Comment by Beuc** [comment]
The nightly must have failed, but I tested with config.gl2=True, with the same results.

At a point the game did start, maybe because I let it wait on the title menu for a while before interacting, but I haven't been able to reproduce it.

LineageOS is akin to installing Windows straight from DVD: the Android OS, sans the vendor's preinstalled stuff. Is one of your devices running Android 9?

**Comment by renpytom** [comment]
I think you're thinking of AOSP. LineageOS is like trying to install Window 10 on a computer using Windows XP drivers, plus some hacks to get them to work - I don't think the result passed the Compatibility Test Suite. 

The nightly did build today, finally, so it will have gl_set_attributes? 

Do you know what model of phone you have? IIRC, there were several variants of the S5. I'm thinking that at this point, I could go hunting ebay for the right one, bring it in, and see what's going on.

**Comment by Beuc** [comment]
I'm curious about what makes you bash the project this way.

Anyway, I upgraded my G900F/klte to Android 10, no changes.

With s/set_gl/gl_set/, with/without the 2 lines commented out, the behavior is also the same (app not responding, occasional black screen and very occasional working run).

:/

**Comment by renpytom** [comment]
There's a bit of frustration from other things, sorry about that. That being said, Lineage OS does run an older kernel with a newer Android, and so has the potential to introduce novel bugs. I worry somewhat that I'm triggering one of those, in which case I don't  think it would be possible to fix.

I have a G900F with a broken screen and G900V coming in, but it'll take a week or two to get here.

**Comment by Beuc** [comment]
OK, thank you for your dedication.
This is the only game/engine that exhibits this behavior so far, so hopefully we'll just find an annoying corner case with the GPU.

**Comment by renpytom** [comment]
So, the phone came in.

I'm pretty sure it should be fixed in 7.4.1.1259. It seems like Ren'Py was destroying and re-creating the GL context unnecessarily, and that was causing problems. I've fixed Ren'Py to not do this. And I have no idea why the GL context being removed like this causes problems, but it was in the offical samsung release and lineageOS.

**Comment by Beuc** [comment]
I tested the latest pre-release and the game runs fine!
GPUs init can be weird :/
Thanks!
